WIP: openjdk: 11.0.2 -> 12

jerith666 opened the pull request:
###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

I still want to run with this for a day or two at work before declaring it okay, but it builds and basically runs.  :)

I don't have a darwin machine so I'll need someone to test that build for me.

We'll probably also want to offload the JDK 11 bootstrap to tarballs.nixos.org.

cc @NeQuissimus @thoughtpolice @MP2E

volth commented:
I would keep 11 for a while.
For example ```ghidra``` [uses jdk11](https://via.hypothes.is/https://ghidra-sre.org/InstallationGuide.html) and might have new issues with ```jdk12```

rycee commented:
This removes OpenJDK 11 entirely? Isn't that the latest LTS version? I think it would be nice to have it available until the next LTS.

jerith666 commented:
It does remove 11 and move all packages that used 11 forward to 12, yeah.  I think this is where we ought to be heading ...

Oracle has been pretty clear in their communications that the recommended path is to move onto each new 6-month release as quickly as possible.  I view the LTS releases as targeted at risk-averse corporate users, who've got the time to invest in big migration projects every 3-4 years (for each LTS release).  I don't think nixpkgs is either that risk-averse or able to make those periodic large investments.  :)

I think the goal ought to be that `pkgs.jdk` & `pkgs.openjdk` refer to the latest release, and that most other packages in nixpgks that use Java should refer to these.  (I've started down that route, but it's tedious and I didn't want it to block getting 12 out there.)

This way everything by default stays on latest.  We can keep the newest LTS (and, sadly, probably 8 as well) around for those who need it, but our goal should be everything in nixpkgs uses the latest.

That said, those are longer term goals; if it's simpler for now to just add 12 alongside 11, and not move any of the dependencies forward in this PR, I'm happy to do that and tackle the rest in separate PRs.

Thanks for the reviews!

volth commented:
Maybe https://github.com/NixOS/nixpkgs/issues/55642 is a better place to collect the opinions (or should we start a RFC?)

jerith666 committed:
openjdk: add 12

* some configure options have been removed upstream

* need a new patch to deal with gcc format warnings

11 remains, as it is an LTS release; all existing users of 11 in
nixpkgs remain on 11 for now.

openjdk/default.nix and openjdk/darwin/default.nix become the
expressions for the current version (12 now; later 13, 14, etc.).

(note: darwin/default.nix was unreferenced; the new version is derived
from darwin/11.nix.)

jerith666 commented:
I haven't done any RFCs yet, will have to read up on that process first.  But for now, I've pushed a more modest version that just adds 12 alongside 11.  :)

rycee commented:
As a Java developer that writes code for risk-averse corporate users I appreciate the availability of Java 11. I'm still mainly using Java 8 😀

I'm ok with the "default packages" (like `jdk` and `openjdk`) to point to the latest version like 12.

@jerith666 Thanks for working on these Java improvements!

rycee commented on pkgs/development/compilers/openjdk/default.nix:
```diff
@@ -0,0 +1,168 @@
+      homepage = http://openjdk.java.net/;
+      license = licenses.gpl2;
+      description = "The open-source Java Development Kit";
+      maintainers = with maintainers; [ edwtjo ];
```
@edwtjo Are you still maintaining the Java packages?

volth commented:
this ought to be ```zulu12``` (like existing ```zulu8```), not ```openjdk12``` ?

symphorien commented on pkgs/development/compilers/openjdk/default.nix:
```diff
@@ -0,0 +1,168 @@
+        done
+      done
+
+      # Test to make sure that we don't depend on the bootstrap
```
If you prefer, you can use `disallowedReferences` instead (see https://nixos.org/nix/manual/#sec-advanced-attributes)

thoughtpolice commented:
I also am of the opinion we should keep the LTS version of Java. Ideally, I think we would always have exactly two versions, and no more:

  - The current latest LTS version, always bumped as quickly as possible when a new major LTS comes out. This would currently be JDK 11 (and in the future, we would move to, uh, JDK 14 I guess, when that happens?)
  - The latest release of the JDK, no matter what it is. In this case, it would currently be JDK 12.

(Whether or not `jdk` and `openjdk` pointed to the first one or second one is a matter of debate, I think, but I believe the above rule could be discussed beyond that.)

The main nice thing about this I think is that the release cadence of Nix and Java isn't always going to align. If the policy is we immediately stop supporting things after 6 months, and the NixOS release cycle is 6 months, then you will often reach situations where you have an EOL version of the JDK mid-release. Then, that version gets no more backports (for example, you may leave a multi-month window open with a security vulnerability). At least with a stable LTS version, users can be certain that they will have a package that won't be EOL'd mid release, and will also exist *between* Nixpkgs releases.

Regarding Oracle's support policy, their new policy is effectively to gate patches to JDK versions after 6 months, which is unfortunate. But the thing is, LTS versions of the JDK generally have *wide* support among the community, for example by Red Hat and IBM. It is very likely for instance that JDK 11 will see many community patches over time, even if Oracle steps away after a set period. So I think we aren't taking a huge risk by keeping them around.

---

There are also other complications with our packaging. For example, what is Zulu doing here? It's really just the Azul-branded/supported version of OpenJDK, there is no actual meaningful technical difference -- but there are Nix packaging distinctions between the two that mean Zulu works in some places OpenJDK does not. Ideally we would fix these issues and get rid of Zulu as well, to keep the overhead lower.

There are also some other things I'd like to see but can't recall off the top of my head right now, like a new update of the bootstrap tarballs (since they're on JDK 10 which has been nuked), but that's another story...

thoughtpolice commented:
And, actually, I would actually also like to even go as far as *removing* Java 8 (gasp) and replacing it everywhere with Java 11 in Nixpkgs. We could keep JDK8 around in something like [NUR](https://github.com/nix-community/NUR) for people who really need it.

There's another benefit to this other than less version maintenance: by doing this, we could also perhaps finally destroy the `oraclejdk` expressions -- for JDK 8, they had feature differences that distinguished them and made the Oracle package a requirement for some things. But, as of JDK 11 there are no actual feature differences between the two, so all uses of OracleJDK in our expressions can hopefully be replaced by OpenJDK, providing we could fix a few packaging issues. This would allow us to remove a bunch of extra crud and get rid of a common proprietary Nix package.

volth commented:
removing ```Java 8``` would mean removal of at least ```hadoop``` and ```processing```

thoughtpolice commented:
A fast google indicates there's some active amount of upstream work on both of those issues. But in general I think the thrust of my point stands: 2 major shipped versions (LTS/latest), and we should make progress towards eliminating the (somewhat absurd) proliferation of JDK/Java packages.

jerith666 commented:
Okay, I've put in three workdays with Eclipse on top of this, no major issues have cropped up.  Has anyone had a chance to test darwin?

volth commented:
For ```zulu${number}``` I would keep everything the upstream maintains and offers on https://www.azul.com/downloads/zulu/ (they just dropped Java 6 and still support Java 7 and Java 8)

jerith666 commented:
Conflict is with https://github.com/NixOS/nixpkgs/pull/63574, so cc @abbradar, @jb55, @risicle, @edwtjo.  Textually it's minor, but conceptually I think I need to re-cut-n-paste `12.nix` to pick up the changes made to `11.nix`.

jerith666 committed:
openjdk11: build fixes

* increase heap size for building javadoc (this is only needed for the
  32-bit build)

* filter out some non-maven jars that crop up in the openjfx build

jerith666 commented:
I've incorporated the OpenJFX work from #63574 now.  Next, I'll update to 12.0.2, which was recently released.

jerith666 committed:
openjdk: 12 -> 12.0.2

gebner commented:
Works for me.  AFAICT all objections have been addressed.  The question of what LTS releases we should keep in nixpkgs should be discussed somewhere else.  I'll merge this later today unless anybody complains.

volth commented:
Build fails here with
```
building '/nix/store/kh94vpfl0bsqcpphc29dmfx12l047vkx-openjfx-deps-12-12+14.drv'...
unpacking sources
unpacking source archive /nix/store/znk95cqc20qwsksgmpacijrwglw65v2f-12+14.tar.gz
source root is rt-12+14
setting SOURCE_DATE_EPOCH to timestamp 1550254664 of file rt-12+14/tools/scripts/whatChanged
patching sources
configuring
no configure script, doing nothing
building

Welcome to Gradle 4.10.3!

Here are the highlights of this release:
 - Incremental Java compilation by default
 - Periodic Gradle caches cleanup
 - Gradle Kotlin DSL 1.0-RC6
 - Nested included builds
 - SNAPSHOT plugin versions in the `plugins {}` block

For more details see https://docs.gradle.org/4.10.3/release-notes.html

WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.codehaus.groovy.reflection.CachedClass (file:/nix/store/mkqi7j178zqp5h1d52niqwrv9jcdqg6z-gradle-4.10.3/lib/gradle/lib/groovy-all-2.4.15.jar) to method java.lang.Object.finalize()
WARNING: Please consider reporting this to the maintainers of org.codehaus.groovy.reflection.CachedClass
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
> Task :buildSrc:compileJava NO-SOURCE

> Task :buildSrc:compileGroovy
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.codehaus.groovy.reflection.CachedClass (file:/nix/store/mkqi7j178zqp5h1d52niqwrv9jcdqg6z-gradle-4.10.3/lib/gradle/lib/groovy-all-2.4.15.jar) to method java.lang.Object.finalize()
WARNING: Please consider reporting this to the maintainers of org.codehaus.groovy.reflection.CachedClass
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release

> Task :buildSrc:processResources NO-SOURCE
> Task :buildSrc:classes
> Task :buildSrc:jar
> Task :buildSrc:assemble
> Task :buildSrc:compileTestJava NO-SOURCE
> Task :buildSrc:compileTestGroovy NO-SOURCE
> Task :buildSrc:processTestResources NO-SOURCE
> Task :buildSrc:testClasses UP-TO-DATE
> Task :buildSrc:test NO-SOURCE
> Task :buildSrc:check UP-TO-DATE
> Task :buildSrc:build

> Configure project :
*****************************************************************
Unsupported gradle version 4.10.3 in use.
Only version 4.8 is supported. Use this version at your own risk
*****************************************************************
gradle.gradleVersion: 4.10.3
OS_NAME: linux
OS_ARCH: amd64
JAVA_HOME: /nix/store/scsd4v46av3y65yf4sxp6i6jwkf191vi-openjdk-11.0.3-ga/lib/openjdk
JDK_HOME: /nix/store/scsd4v46av3y65yf4sxp6i6jwkf191vi-openjdk-11.0.3-ga/lib/openjdk
java.runtime.version: 11.0.3-internal+0-adhoc..jdk11u-jdk-11.0.3-ga
java version: 11.0.3
java build number: 0
jdk.runtime.version: 11.0.3-internal+0-adhoc..jdk11u-jdk-11.0.3-ga
jdk version: 11.0.3
jdk build number: 0
minimum jdk version: 11
minimum jdk build number: 28
HAS_JAVAFX_MODULES: true
STUB_RUNTIME: /nix/store/scsd4v46av3y65yf4sxp6i6jwkf191vi-openjdk-11.0.3-ga/lib/openjdk
CONF: Release
NUM_COMPILE_THREADS: 12
COMPILE_TARGETS: linux
COMPILE_FLAGS_FILES: buildSrc/linux.gradle
HUDSON_JOB_NAME: not_hudson
HUDSON_BUILD_NUMBER: 0000
PROMOTED_BUILD_NUMBER: 0
PRODUCT_NAME: OpenJFX
RELEASE_VERSION: 12
RELEASE_SUFFIX: -internal
RELEASE_VERSION_SHORT: 12-internal
RELEASE_VERSION_LONG: 12-internal+0-2019-08-06-091642
RELEASE_VERSION_PADDED: 12.0.0.0
MAVEN_VERSION: 12-internal+0-2019-08-06-091642
UPDATE_STUB_CACHE: false
Building Webkit configuration /Release/ into /build/rt-12+14/modules/javafx.web/build/linux
module: project ':apps' (buildModule=NO)
module: project ':base' (buildModule=YES)
module: project ':controls' (buildModule=YES)
module: project ':fxml' (buildModule=YES)
module: project ':graphics' (buildModule=YES)
module: project ':media' (buildModule=YES)
module: project ':swing' (buildModule=YES)
module: project ':swt' (buildModule=NO)
module: project ':systemTests' (buildModule=NO)
module: project ':web' (buildModule=YES)

FAILURE: Build failed with an exception.

* Where:
Build file '/build/rt-12+14/build.gradle' line: 3928

* What went wrong:
A problem occurred evaluating root project 'rt-12+14'.
> Could not resolve all files for configuration ':apps:lucene'.
   > Could not download lucene-core.jar (org.apache.lucene:lucene-core:7.4.0)
      > Could not get resource 'https://repo.maven.apache.org/maven2/org/apache/lucene/lucene-core/7.4.0/lucene-core-7.4.0.jar'.
         > Read timed out

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 5.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/4.10.3/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 57s
note: keeping build directory '/tmp/nix-build-openjfx-deps-12-12+14.drv-0'
builder for '/nix/store/kh94vpfl0bsqcpphc29dmfx12l047vkx-openjfx-deps-12-12+14.drv' failed with exit code 1
```

volth commented:
@GrahamcOfBorg build openjdk12

gebner commented:
@volth The build works on ofborg as well (it hits the point where you got the error before the timeout).  I'll merge it, let's see what hydra says.  Even if it fails, it shouldn't break anything since nothing depends on openjdk12 yet.
